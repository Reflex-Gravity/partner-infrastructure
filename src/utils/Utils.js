export default class Utils {

  static filterArrayByString(mainArr, searchKey, searchText)
    {
        if ( searchText === '' )
        {
            return mainArr;
        }

        searchText = searchText.toLowerCase();
        //  return filtered array.
        return mainArr.filter(itemObj => {
            return this.searchInObj(itemObj[searchKey], searchText);
        });
    };

    static searchInObj(objValue, searchText)
    {
      if (this.searchInString(objValue, searchText))
      {
          return true;
      }

    }

    static searchInString(value, searchText)
    {
        return value.toLowerCase().includes(searchText);
    }
}