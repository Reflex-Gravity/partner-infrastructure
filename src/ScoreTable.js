/* eslint-disable no-script-url */

import React, {useEffect} from 'react';
import Link from '@material-ui/core/Link';
import { makeStyles } from '@material-ui/core/styles';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Title from './Title';
import { TableSortLabel, Input, Select, MenuItem } from '@material-ui/core';
import Utils from './utils/Utils'

const useStyles = makeStyles(theme => ({
  seeMore: {
    marginTop: theme.spacing(3),
  },
  searchBox: {
    paddingLeft : theme.spacing(1),
    marginBottom : theme.spacing(3),
    marginTop : theme.spacing(3),
  },
  input: {
    margin: theme.spacing(1),
  },
}));

function ScoreTable( props ) {
  const classes = useStyles();
  const [orderType, setOrderType] = React.useState('asc');
  const [orderField, setOrderField] = React.useState('');
  const [showMore, setShowMore] = React.useState(false);
  const [searchType, setSearchType] = React.useState('last_name');
  const [peopleData, setPeopleData] = React.useState([]);

  useEffect(() => {
    // If show more is clicked, then display all data.
    if (showMore === true) {
      setPeopleData(props.peopleData)

    }
  }, [showMore, props.peopleData]);

  useEffect(() => {
    // Initial mount/render show only 10 items.
    setPeopleData([...props.peopleData].splice(0, 10))
  }, [props.peopleData]);


  const sortData = (sortField ) => (evt) => {
    setOrderField(sortField);
    
    setOrderType( orderType == 'asc' ? 'desc' : 'asc');
    let sortedPeopleData = [...peopleData];

    // Sort the data based on field.
    sortedPeopleData.sort(( a, b)=>{
        
        if( b[sortField] < a[sortField] ){
          // If earlier sorting/order type was ascending then, order by descending, vice-versa.
          return orderType == 'asc' ? -1 : 1;
        }
        
        if( b[sortField] > a[sortField]){
          // If earlier sorting/order type was ascending then, order by descending, vice-versa.
          return orderType == 'asc' ? 1 : -1;
        }

      return 0;
    });

    setPeopleData(sortedPeopleData);
  };

  const searchData = (evt) => {
    // Filter the data based on search
    setPeopleData(Utils.filterArrayByString([...props.peopleData], searchType, evt.target.value))
  };

  const showMoreData = () =>{
    setShowMore(true);
  };

  const handleSearchTypeChange = (evt)=>{
    setSearchType(evt.target.value);
  };

  return (
    <React.Fragment>
      <Title>Scores listing</Title>
      <div className={classes.searchBox}>
        <Select
          value={searchType}
          onChange={handleSearchTypeChange}
        >
          <MenuItem value="last_name">Last Name</MenuItem>
          <MenuItem value="first_name">First Name</MenuItem>
          <MenuItem value="country">Country</MenuItem>
        </Select>
        <Input
          placeholder="Search listing..."
          className="searchField"
          onChange={searchData}
          inputProps={{
            'aria-label': 'Search the scores',
          }}
        />
      </div>
      <Table size="small">
        <TableHead>
          <TableRow>
            <TableCell>
              <TableSortLabel
                active = {orderField === "last_name"}
                direction={orderType}
                onClick={sortData( "last_name" )}
              >
                Last name
              </TableSortLabel>
            </TableCell>
            <TableCell>
              <TableSortLabel
                active = {orderField === "first_name"}
                direction={orderType}
                onClick={sortData( "first_name" )}
              >
                First name
              </TableSortLabel>
            </TableCell>
            <TableCell>Gender</TableCell>
            <TableCell>City</TableCell>
            <TableCell>
              <TableSortLabel
                className="country-head"
                active = {orderField === "country"}
                direction={orderType}
                onClick={sortData( "country" )}
              >
                Country
              </TableSortLabel>
            </TableCell>
            <TableCell align="right">
              Score
            </TableCell>
          </TableRow>
        </TableHead>
        <TableBody className="scoreTableBody">
          {peopleData.map(row => (
            <TableRow key={row.id}>
              <TableCell>{row.last_name}</TableCell>
              <TableCell>{row.first_name}</TableCell>
              <TableCell>{row.gender}</TableCell>
              <TableCell>{row.city}</TableCell>
              <TableCell>{row.country}</TableCell>
              <TableCell align="right">{row.score}</TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
      <div className={classes.seeMore}>
        <Link color="primary" onClick={showMoreData} href="#">
          See more scores
        </Link>
      </div>
    </React.Fragment>
  );
}

export default React.memo(ScoreTable)