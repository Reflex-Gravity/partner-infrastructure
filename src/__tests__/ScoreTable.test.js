import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer'
import Enzyme, { shallow, render, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16'

import Utils from '../utils/Utils'
import ScoreTable from '../ScoreTable';
import {peopleData} from '../peopleData'

Enzyme.configure({adapter: new Adapter()})

it(' renders countryBased component correctly', ()=>{
  const tree = renderer
                .create(
                  <ScoreTable peopleData={peopleData}/>
                )
                .toJSON();
      expect(tree).toMatchSnapshot();
});

test('check search filter', ()=>{

  const search = Utils.filterArrayByString([...peopleData], "first_name", "murdock");
  
  expect(search).toEqual([{
    "id": "a67e6828-99d9-4d8d-9cd7-8aff12e95973",
    "first_name": "Murdock",
    "last_name": "Ledstone",
    "email": "mledstone0@mayoclinic.com",
    "gender": "Male",
    "city": "Olofstr√∂m",
    "country": "SE",
    "score": null,
    "created_at": "2017-04-05T02:28:37Z"
  }])

})