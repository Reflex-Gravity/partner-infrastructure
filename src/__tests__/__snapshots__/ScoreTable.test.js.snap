// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` renders countryBased component correctly 1`] = `
Array [
  <h2
    className="MuiTypography-root MuiTypography-h6 MuiTypography-colorPrimary MuiTypography-gutterBottom"
  >
    Scores listing
  </h2>,
  <div
    className="makeStyles-searchBox-2"
  >
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline"
      onClick={[Function]}
    >
      <div
        aria-haspopup="listbox"
        className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input MuiInputBase-inputSelect"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        role="button"
        tabIndex={0}
      >
        Last Name
      </div>
      <input
        type="hidden"
        value="last_name"
      />
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root MuiSelect-icon"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline searchField"
      onClick={[Function]}
    >
      <input
        aria-label="Search the scores"
        className="MuiInputBase-input MuiInput-input"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        placeholder="Search listing..."
        type="text"
      />
    </div>
  </div>,
  <table
    className="MuiTable-root"
  >
    <thead
      className="MuiTableHead-root"
    >
      <tr
        className="MuiTableRow-root MuiTableRow-head"
      >
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
          scope="col"
        >
          <span
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Last name
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
          scope="col"
        >
          <span
            className="MuiButtonBase-root MuiTableSortLabel-root"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            First name
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
          scope="col"
        >
          Gender
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
          scope="col"
        >
          City
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeSmall"
          scope="col"
        >
          <span
            className="MuiButtonBase-root MuiTableSortLabel-root country-head"
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            role="button"
            tabIndex={0}
          >
            Country
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
        <th
          aria-sort={null}
          className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeSmall"
          scope="col"
        >
          Score
        </th>
      </tr>
    </thead>
    <tbody
      className="MuiTableBody-root scoreTableBody"
    />
  </table>,
  <div
    className="makeStyles-seeMore-1"
  >
    <a
      className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
      href="#"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
    >
      See more scores
    </a>
  </div>,
]
`;
